examples/multi_gpu/papers100m_gcn_cugraph.py

```diff
@@ -0,0 +1,330 @@
+    return client, cluster
+
+
+def shutdown_dask_client(client):
+    from cugraph.dask.comms import comms as Comms
+    Comms.destroy()
+    client.close()
+
+
+def pyg_num_work(world_size):
+    num_work = None
+    if hasattr(os, "sched_getaffinity"):
+        try:
+            num_work = len(os.sched_getaffinity(0)) / (2 * world_size)
+        except Exception:
+            pass
+    if num_work is None:
+        num_work = os.cpu_count() / (2 * world_size)
+    return int(num_work)
+
+
+def init_pytorch_worker(rank, world_size):
+    import rmm
+    if rank > 0:
+        rmm.reinitialize(devices=rank)
+
+    import cupy
+    cupy.cuda.Device(rank).use()
+    from rmm.allocators.cupy import rmm_cupy_allocator
+    cupy.cuda.set_allocator(rmm_cupy_allocator)
+
+    from cugraph.testing.mg_utils import enable_spilling
+    enable_spilling()
+
+    torch.cuda.set_device(rank)
+
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '12355'
+    dist.init_process_group('nccl', rank=rank, world_size=world_size)
+
+
+def run_train(rank, data, world_size, model, epochs, batch_size, fan_out,
+              split_idx, num_classes, wall_clock_start, tempdir=None,
+              num_layers=3):
+
+    init_pytorch_worker(
+        rank,
+        world_size,
+    )
+
+    if rank == 0:
+        client, cluster = start_dask_cluster()
+        from cugraph.dask.comms import comms as Comms
+        Comms.initialize(p2p=True)
+    model = model.to(rank)
+    model = DistributedDataParallel(model, device_ids=[rank])
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.01,
+                                 weight_decay=0.0005)
+
+    kwargs = dict(
+        num_neighbors=[fan_out] * num_layers,
+        batch_size=batch_size,
+    )
+    # Set Up Neighbor Loading
+    import cugraph
+    from cugraph_pyg.data import CuGraphStore
+    from cugraph_pyg.loader import BulkSampleLoader
+    G = {("N", "E", "N"): data.edge_index}
+    N = {"N": data.num_nodes}
+    fs = cugraph.gnn.FeatureStore(backend="torch")
+    fs.add_data(data.x, "N", "x")
+    fs.add_data(data.y, "N", "y")
+    dist.barrier()
+
+    if rank == 0:
+        print("Rank 0 creating its cugraph store and \
+            initializing distributed graph")
+        cugraph_store = CuGraphStore(fs, G, N, multi_gpu=True)
+        print("Distributed graph initialization complete.")
+
+    if rank != 0:
+        print(f"Rank {rank} waiting for distributed graph initialization")
+    dist.barrier()
+
+    if rank != 0:
+        print(f"Rank {rank} proceeding with store creation")
+        cugraph_store = CuGraphStore(fs, {
+            k: len(v)
+            for k, v in G.items()
+        }, N, multi_gpu=False)
+        print(f"Rank {rank} created store")
+    dist.barrier()
+
+    if rank == 0:
+        # Direct cuGraph to sample offline prior to the training loop
+        # Sampling will occur in parallel but will be initiated on rank 0
```

Comment:
Is that necessary?

Reply:
currently yes. this is due to limitations with Dask. CuGraph team is trying to remove dependencies on dask. in the future this wont be an issue. we will update example at that time.